tests/langchain/test_langchain_autolog.py

```diff
@@ -770,284 +781,270 @@ async def on_chain_end(self, outputs: Dict[str, Any], **kwargs: Any) -> None:
         self.logs.append("chain_end")
 
 
+_CONFIG_PATTERNS = [
+    # Config with no user callbacks
+    RunnableConfig(max_concurrency=1),
+    # With user callbacks
+    RunnableConfig(callbacks=[CustomCallbackHandler()]),
+    RunnableConfig(callbacks=BaseCallbackManager([CustomCallbackHandler()])),
+]
+
+_ASYNC_CONFIG_PATTERNS = [
+    RunnableConfig(callbacks=[AsyncCustomCallbackHandler()]),
+    RunnableConfig(callbacks=BaseCallbackManager([AsyncCustomCallbackHandler()])),
+]
+
+
+def _extract_callback_handlers(config) -> Optional[List[BaseCallbackHandler]]:
+    if isinstance(config, list):
+        callbacks = []
+        for c in config:
+            callbacks_in_c = _extract_callback_handlers(c)
+            if callbacks_in_c:
+                callbacks.extend(callbacks_in_c)
+        return callbacks
+    # RunnableConfig is also a dict
+    elif isinstance(config, dict) and "callbacks" in config:
+        callbacks = config["callbacks"]
+        if isinstance(callbacks, BaseCallbackManager):
+            return callbacks.handlers
+        else:
+            return callbacks
+    else:
+        return None
+
+
 @pytest.mark.parametrize("invoke_arg", ["args", "kwargs", None])
-@pytest.mark.parametrize(
-    "generate_callbacks",
-    [
-        lambda: [CustomCallbackHandler()],
-        lambda: BaseCallbackManager([CustomCallbackHandler()]),
-        lambda: None,
-    ],
-)
-def test_langchain_autolog_callback_injection_in_invoke(invoke_arg, generate_callbacks):
-    mlflow.langchain.autolog(log_models=True, extra_tags={"test_tag": "test"})
-    callbacks = generate_callbacks()
-    with mlflow.start_run() as run, _mock_request(return_value=_mock_chat_completion_response()):
-        model = create_openai_llmchain()
+@pytest.mark.parametrize("config", _CONFIG_PATTERNS)
+def test_langchain_autolog_callback_injection_in_invoke(invoke_arg, config):
+    mlflow.langchain.autolog()
+
+    model = create_openai_runnable()
+    original_handlers = _extract_callback_handlers(config)
+
+    with _mock_request(return_value=_mock_chat_completion_response()):
+        input = {"product": "MLflow"}
         if invoke_arg == "args":
-            model.invoke("MLflow", RunnableConfig(callbacks=callbacks))
+            model.invoke(input, config)
         elif invoke_arg == "kwargs":
-            model.invoke("MLflow", config=RunnableConfig(callbacks=callbacks))
+            model.invoke(input, config=config)
         elif invoke_arg is None:
-            model.invoke("MLflow")
-        assert mlflow.active_run() is not None
-    run_data = MlflowClient().get_run(run.info.run_id).data
-    assert run_data.tags["test_tag"] == "test"
```

Comment:
While I agree this should be removed from the focus of this unit test, can we ensure that tag inclusion functionality for autologging has coverage in a followup if required?

Reply:
Tag functionality is covered in the other tests: https://github.com/mlflow/mlflow/blob/bda5dd58a3f39aec59be72ff26ee431ce818bb72/tests/langchain/test_langchain_autolog.py#L193-L194

Reply:
Perfect!